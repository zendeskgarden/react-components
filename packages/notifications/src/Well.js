/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import PropTypes from 'prop-types';
import styled from 'styled-components';
import classNames from 'classnames';
import CalloutStyles from '@zendeskgarden/css-callouts';
import { retrieveTheme, isRtl } from '@zendeskgarden/react-theming';

const COMPONENT_ID = 'notifications.well';

/**
 * Supports all `<div>` props
 */
const Well = styled.div.attrs(props => ({
  'data-garden-id': COMPONENT_ID,
  'data-garden-version': PACKAGE_VERSION,
  className: classNames(CalloutStyles['c-callout'], {
    // RTL
    [CalloutStyles['is-rtl']]: isRtl(props),

    // Styles
    [CalloutStyles['c-callout--recessed']]: props.recessed,
    [CalloutStyles['c-callout--dialog']]: props.floating
  })
}))`
  ${props => retrieveTheme(COMPONENT_ID, props)};
`;

Well.propTypes = {
  recessed: PropTypes.bool,
  floating: PropTypes.bool
};

/** @component */
export default Well;
