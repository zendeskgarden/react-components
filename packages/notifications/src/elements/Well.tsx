/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { HTMLAttributes } from 'react';
import PropTypes from 'prop-types';
import { StyledWell } from '../styled';

interface IWellProps {
  /** Applies a background color */
  isRecessed?: boolean;
  /** Applies a drop shadow */
  isFloating?: boolean;
}

/**
 * @extends HTMLAttributes<HTMLDivElement>
 */
export const Well = React.forwardRef<HTMLDivElement, IWellProps & HTMLAttributes<HTMLDivElement>>(
  (props, ref) => <StyledWell ref={ref} {...props} />
);

Well.displayName = 'Well';

Well.propTypes = {
  isRecessed: PropTypes.bool,
  isFloating: PropTypes.bool
};
