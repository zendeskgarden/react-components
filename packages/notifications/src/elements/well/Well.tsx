/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { IWellProps } from '../../types';
import { StyledWell } from '../../styled';
import { Title } from '../Title';
import { Paragraph } from '../Paragraph';

export const WellComponent = React.forwardRef<HTMLDivElement, IWellProps>(
  ({ isFloating, isRecessed, ...props }, ref) => (
    <StyledWell ref={ref} $isFloating={isFloating} $isRecessed={isRecessed} {...props} />
  )
);

WellComponent.displayName = 'Well';

WellComponent.propTypes = {
  isRecessed: PropTypes.bool,
  isFloating: PropTypes.bool
};

/**
 * @extends HTMLAttributes<HTMLDivElement>
 */
export const Well = WellComponent as typeof WellComponent & {
  Paragraph: typeof Paragraph;
  Title: typeof Title;
};

Well.Paragraph = Paragraph;
Well.Title = Title;
