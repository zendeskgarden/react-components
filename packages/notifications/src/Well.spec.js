/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React from 'react';
import { render, renderRtl } from 'garden-test-utils';
import Well from './Well';

describe('Well', () => {
  it('renders default well styling', () => {
    const { container } = render(<Well />);

    expect(container.firstChild).toHaveClass('c-callout');
  });

  it('renders with RTL styling if applied', () => {
    const { container } = renderRtl(<Well />);

    expect(container.firstChild).toHaveClass('is-rtl');
  });

  it('renders recessed styling correctly', () => {
    const { container } = render(<Well recessed />);

    expect(container.firstChild).toHaveClass('c-callout--recessed');
  });

  it('renders floating styling correctly', () => {
    const { container } = render(<Well floating />);

    expect(container.firstChild).toHaveClass('c-callout--dialog');
  });
});
