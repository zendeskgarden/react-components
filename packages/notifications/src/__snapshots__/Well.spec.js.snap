// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Well renders default well styling 1`] = `
<div
  className="c-callout "
  data-garden-id="notifications.well"
  data-garden-version="version"
/>
`;

exports[`Well renders floating styling correctly 1`] = `
<div
  className="c-callout c-callout--dialog "
  data-garden-id="notifications.well"
  data-garden-version="version"
/>
`;

exports[`Well renders recessed styling correctly 1`] = `
<div
  className="c-callout c-callout--recessed "
  data-garden-id="notifications.well"
  data-garden-version="version"
/>
`;

exports[`Well renders with RTL styling if applied 1`] = `
<div
  className="c-callout is-rtl "
  data-garden-id="notifications.well"
  data-garden-version="version"
/>
`;
